### ISSUE TITLE
rust-analyzer doesn't fully respect CARGO_TARGET_DIR set with rust-analyzer.server.extraEnv

### ISSUE THREAD

**Issue by willhodges** [rust-lang/rust-analyzer#12760]
Rust-analyzer seems to kick off a cargo command in such a way that a CARGO_TARGET_DIR override set with rust-analyzer.server.extraEnv set in rust-analyzer.json in the project root. This issue popped up in trying to address issue #5828 , where rust-analyzer is invalidating my cargo cache during code analysis in a Cargo workspace with sub-crates that have different build targets.

I've made a fairly small repro case to demonstrate the issue at https://github.com/willhodges/rabug. Simply clone into the repo and then open it in VSCode (with the rust-analyzer extension) or Atom (with the ide-rust extension). The target/debug directory is immediately populated during the first rust-analyzer run despite an override of CARGO_TARGET_DIR in the project's .cargo/config.

**rust-analyzer version**: rust-analyzer 0.0.0 (5342f47f4 2022-07-09)

**rustc version**: rustc 1.62.0 (a8314ef7d 2022-06-27)

**relevant settings**: See .cargo/config.toml and rust-analyzer.json in the repository above

**Comment by willhodges** [comment]
@flodiebold 

> I don't know how the configuration works in Atom, but VSCode isn't going to respect settings in a `rust-analyzer.json`.

This doesn't appear to be correct. If you pull down the project and open it in VSCode, rust-analyzer runs and target/check is populated. The only place that dir is specified is in rust-analyzer.json.

> Also, server.extraEnv is a setting specific to the VSCode extension. Other extensions might have different ways to configure the environment.

I'm fine with using VSCode for the repro, since it seems to be one of the more popular editors for Rust.

**Comment by flodiebold** [comment]
> This doesn't appear to be correct. If you pull down the project and open it in VSCode, rust-analyzer runs and target/check is populated. The only place that dir is specified is in rust-analyzer.json. target/debug is also populated, which suggests the environment variable is not respected somewhere along the way.

That's not the case for me. Maybe you have the target dir configured somewhere else as well? Do you have any global VSCode settings for rust-analyzer?

**Comment by willhodges** [comment]
I don't believe so. I can try to test this in a clean VM later this weekend.

**Comment by flodiebold** [comment]
You could also try changing the path configured in `rust-analyzer.json` to something else (e.g. `target/check2`) and see what happens.
